shaders/ray_tracing_reflection/closesthit.rchit

```diff
@@ -0,0 +1,145 @@
+/*
+ * Copyright (c) 2021, NVIDIA CORPORATION.  All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-FileCopyrightText: Copyright (c) 2014-2021 NVIDIA CORPORATION
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#version 460
+#extension GL_EXT_ray_tracing : enable
+#extension GL_EXT_scalar_block_layout : enable
+#extension GL_EXT_nonuniform_qualifier : enable
+
+struct hitPayload
+{
+	vec3 radiance;
+	vec3 attenuation;
+	int  done;
+	vec3 rayOrigin;
+	vec3 rayDir;
+};
+
+layout(location = 0) rayPayloadInEXT hitPayload prd;
+layout(location = 1) rayPayloadEXT bool isShadowed;
+
+hitAttributeEXT vec3 attribs;
+
+struct WaveFrontMaterial
+{
+	vec3  diffuse;
+	vec3  specular;
+	float shininess;
+};
+
+struct Vertex
+{
+	vec3 pos;
+	vec3 nrm;
+};
+
+// clang-format off
+layout(set = 0, binding = 0) uniform accelerationStructureEXT topLevelAS;
+layout(set = 0, binding = 3, scalar) buffer _materials { WaveFrontMaterial m[]; } materials[];
+layout(set = 0, binding = 4, scalar) buffer _vertices { Vertex v[]; } vertices[];
+layout(set = 0, binding = 5)         buffer _indices { uint i[]; } indices[];
+layout(set = 0, binding = 6)         buffer _mat_index { int i[]; } mat_index[];
```

Comment:
I believe the above bindings would be OpTypeRuntimeArray, as they are unsized. Ref. VUID-StandaloneSpirv-OpTypeRuntimeArray-04680, I don't belive that is allowed except for the _last_ binding in a set.

In practice, as the bindings are not unsized at API level (they layout use `obj_models.size()`), that's probably why this works fine in practice. But as far as I can tell it's not valid usage.

Reply:
Hm, I'm misreading the spec here. The OpTypeRuntimeArray is last in the structs -- so this is fine. What I was looking for was confirmation that it's OK/not OK to have the descriptor counts be unsized in the SPV, _as long as_ they are sized at API level (which they are in this case). 

What I have in mind is that if they are unsized at API level, the binding must be _last_ in the set -- not the case here -- but this case is not that; this is just dropping the array size in the SPV (but still providing them at API level). So far I can't see anything saying this is not valid usage -- so presumably it's OK? :)

Reply:
Right, so reading up a bit, the key might appear to be that this was invalid prior to SPV_EXT_descriptor_indexing and co, which adds (quoting SPV_EXT_descriptor_indexing):

> The RuntimeDescriptorArrayEXT capability is used to indicate that a module uses arrays of resources which are declared with OpTypeRuntimeArray.

... and (from VK_EXT_descriptor_indexing):

> The final binding in a descriptor set layout can have a variable size (and unsized arrays of resources are allowed in the GL_EXT_nonuniform_qualifier and SPV_EXT_descriptor_indexing extensions).

This seems to allow what is needed. Note that VK_DESCRIPTOR_BINDING_VARIABLE_DESCRIPTOR_COUNT_BIT is _not_ needed, as the size is specified at layout creation time, not descriptor set allocation time.

I will say, it's not totally clear to me what language invalidates VUID-StandaloneSpirv-OpTypeRuntimeArray-04680 as seen in the 1.2 spec:
> OpTypeRuntimeArray **must only** be used for the last member of an OpTypeStruct that is in the StorageBuffer or PhysicalStorageBuffer storage class decorated as Block, or that is in the Uniform storage class decorated as BufferBlock

But the 1.2 spec _also_ says:
> OpTypeRuntimeArray **must only** be used for an array of variables with storage class Uniform, StorageBuffer, or UniformConstant, or for the outermost dimension of an array of arrays of such variables if the runtimeDescriptorArray feature is enabled,

Maybe I'm missing something, but these two statements seems to not agree? Either way, by the latter the use case here seems OK (assuming runtimeDescriptorArray  is enabled of course -- which it explicitly is in the code here)

Reply:
Thanks for looking more deeply into this. I have double check with a colleague (Jeff Bolz), co-author of the extension, and he find no issue with the current usage. Maybe the specification could be made more clear. 

Another way to achieve the same thing, would be to use [buffer_reference](https://github.com/KhronosGroup/GLSL/blob/master/extensions/ext/GLSL_EXT_buffer_reference.txt), but it will introduce an extension that wasn't use in the existing samples.